polyfill/lib/datetime.mjs

```diff
@@ -150,17 +198,71 @@ export class DateTime {
       disambiguation
     ));
     const Construct = ES.SpeciesConstructor(this, DateTime);
-    const result = new Construct(year, month, day, hour, minute, second, millisecond, microsecond, nanosecond);
+    const result = new Construct(
+      year,
+      month,
+      day,
+      hour,
+      minute,
+      second,
+      millisecond,
+      microsecond,
+      nanosecond,
+      calendar
+    );
+    if (!ES.IsTemporalDateTime(result)) throw new TypeError('invalid result');
+    return result;
+  }
+  withCalendar(calendar) {
+    if (!ES.IsTemporalDateTime(this)) throw new TypeError('invalid receiver');
+    const TemporalCalendar = GetIntrinsic('%Temporal.Calendar%');
+    calendar = TemporalCalendar.from(calendar);
+    const Construct = ES.SpeciesConstructor(this, DateTime);
+    const result = new Construct(
+      GetSlot(this, ISO_YEAR),
+      GetSlot(this, ISO_MONTH),
+      GetSlot(this, ISO_DAY),
+      GetSlot(this, HOUR),
+      GetSlot(this, MINUTE),
+      GetSlot(this, SECOND),
+      GetSlot(this, MILLISECOND),
+      GetSlot(this, MICROSECOND),
+      GetSlot(this, NANOSECOND),
+      calendar
+    );
     if (!ES.IsTemporalDateTime(result)) throw new TypeError('invalid result');
     return result;
   }
   plus(temporalDurationLike, options) {
     if (!ES.IsTemporalDateTime(this)) throw new TypeError('invalid receiver');
     const disambiguation = ES.ToTemporalDisambiguation(options);
     const duration = ES.ToLimitedTemporalDuration(temporalDurationLike);
-    let { year, month, day, hour, minute, second, millisecond, microsecond, nanosecond } = this;
-    let { years, months, weeks, days, hours, minutes, seconds, milliseconds, microseconds, nanoseconds } = duration;
-    ({ year, month, day } = ES.AddDate(year, month, day, years, months, weeks, days, disambiguation));
+
+    // Delegate the date part addition to the calendar
+    const calendar = GetSlot(this, CALENDAR);
+    const Date = GetIntrinsic('%Temporal.Date%');
+    const addedDate = calendar.plus(
+      new Date(GetSlot(this, ISO_YEAR), GetSlot(this, ISO_MONTH), GetSlot(this, ISO_DAY), calendar),
+      duration,
+      options,
+      Date
+    );
+    let year = GetSlot(addedDate, ISO_YEAR);
+    let month = GetSlot(addedDate, ISO_MONTH);
+    let day = GetSlot(addedDate, ISO_DAY);
+
+    // Add the time part
+    let { hour, minute, second, millisecond, microsecond, nanosecond } = this;
```

Comment:
The observable calls aren't in alphabetical order if you do this I think (here and below).